corehq/apps/user_importer/helpers.py

```diff
@@ -9,3 +16,283 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+                users_updated_primary_location_name = get_user_primary_location_name(self.user, self.user_domain)
+                self.logger.add_info(
+                    _(f"Primary location: {users_updated_primary_location_name}[{location_id}]"))
+        else:
+            self.user.unset_location(self.user_domain)
+            # if there was a location before, log that it was cleared
+            if current_primary_location_id:
+                self.logger.add_info("Primary location: None")
+
+    def update_locations(self, location_codes, membership, domain_info):
+        from corehq.apps.user_importer.importer import (
+            check_modified_user_loc,
+            find_location_id,
+            get_location_from_site_code
+        )
+
+        location_ids = find_location_id(location_codes, domain_info.location_cache)
+        users_current_primary_location_id = membership.location_id
+        locations_updated, primary_loc_removed = check_modified_user_loc(location_ids,
+                                                                         membership.location_id,
+                                                                         membership.assigned_location_ids)
+        if primary_loc_removed:
+            self.user.unset_location(self.user_domain, commit=False)
+        if locations_updated:
+            self.user.reset_locations(self.user_domain, location_ids, commit=False)
+            if location_ids:
+                locations = [get_location_from_site_code(code, domain_info.location_cache)
+                             for code in location_codes]
+                locations_info = ", ".join([f"{location.name}[{location.location_id}]" for location in locations])
+            else:
+                locations_info = []
+            self.logger.add_info(_(f"Assigned locations: {locations_info}"))
+
+        # log this after assigned locations are updated, which can re-set primary location
+        users_updated_primary_location_id = get_user_primary_location_id(self.user, self.user_domain)
+        if users_updated_primary_location_id != users_current_primary_location_id:
+            if users_updated_primary_location_id:
+                users_updated_primary_location_name = get_user_primary_location_name(self.user, self.user_domain)
+                location_info = f"{users_updated_primary_location_name}[{users_updated_primary_location_id}]"
+            else:
+                location_info = None
+            self.logger.add_info(
+                _(f"Primary location: {location_info}"))
+
+    def save(self):
+        # Tracking for role is done post user save to have role setup correctly on save
+        if self.role_updated:
+            new_role = self.user.get_role(domain=self.user_domain)
+            if new_role:
+                self.logger.add_info(_(f"Role: {new_role.name}[{new_role.get_id}]"))
+            else:
+                self.logger.add_change_message("Role: None")
+        self.logger.save()
```

Comment:
Seems identical to `CommCareUserImporter.save` - maybe put it in a base class?

Reply:
https://github.com/dimagi/commcare-hq/pull/29935/commits/a884b422896976e6797e704d551dfc27ca3fb5c3